File: ToDo.Core/TaskController.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Threading.Tasks;
using ToDo.API;

namespace ToDo.Core
{
    public class TaskController : ITaskController
    {
        private readonly ITaskStorage _taskStorage;
        private long? _newIdForUse = null;

        /// <summary>
        /// Creates object of Task Controller
        /// </summary>
        /// <param name="taskStorage">Storage to be used to preserve tasks' states</param>
        public TaskController(ITaskStorage taskStorage)
        {
            _taskStorage = taskStorage;
        }

        public async Task<IToDoTask> AddTask(string description)
        {
            if (string.IsNullOrWhiteSpace(description))
            {
                return null;
            }

            if (_newIdForUse == null)
            {
                _newIdForUse = await _taskStorage.GetTotalNumberOfTasks();
            }
            return await _taskStorage.AddNewTask((long)_newIdForUse++, description, CreateToDoTask);
        }

        public async Task<IReadOnlyDictionary<long, IToDoTask>> GetAllTasks()
        {
            var tasks = await _taskStorage.GetAllTasks(CreateToDoTask);
            foreach (var task in tasks)
            {
                ((ToDoTask)task.Value).PropertyChanged += OnTaskAttributesChanged;
            }

            return tasks;
        }

        private void OnTaskAttributesChanged(object sender, PropertyChangedEventArgs e)
        {
            ToDoTask toDoTask = (ToDoTask)sender;
            switch (e.PropertyName)
            {
                case nameof(ToDoTask.Description):
                    _taskStorage.UpdateTaskDescription(toDoTask.Id, toDoTask.Description);
                    return;
                case nameof(ToDoTask.HasCompleted):
                    _taskStorage.UpdateTaskCompleteness(toDoTask.Id, toDoTask.HasCompleted);
                    return;
                default:
                    Debug.Fail("We should update storage for this property as well");
                    return;
            }
        }

        private IToDoTask CreateToDoTask(long id, string description, bool hasCompleted)
        {
            var retrivedTask = new ToDoTask(id, description, hasCompleted);
            retrivedTask.PropertyChanged += OnTaskAttributesChanged;
            return retrivedTask;
        }
    }
}

File: ToDo.Core/TaskController.cpp
#include "pch.h"
#include <future>

#include "TaskController.h"
#include "ToDoTask.h"

namespace ToDo::Core {
API::IToDoTask &CreateTask(long id, std::string description,
                           bool hasCompleted) {
  return *(new ToDoTask(id, description, hasCompleted));
}

TaskController::TaskController(API::ITaskStorage &taskStorage) {
  _taskStorage = &taskStorage;
}

void TaskController::AddTask(
    std::string description,
    std::function<void(std::optional<std::reference_wrapper<API::IToDoTask>>)>
        callback) {
  std::async(std::launch::async, [&]() {
    if (description.empty()) {
      return callback(std::nullopt);
    }

    if (!_newIdForUse) {
      _newIdForUse = _taskStorage->GetTotalNumberOfTasks();
    }

    callback(std::optional<std::reference_wrapper<API::IToDoTask>>{
        _taskStorage->AddNewTask((*_newIdForUse)++, description, CreateTask)});
  });
}

void TaskController::OnTaskAttributesChanged(ToDoTask &sender,
                                             TaskAttribute changedAttribute) {
  switch (changedAttribute) {
  case TaskAttribute::Description:
    _taskStorage->UpdateTaskDescription(sender.GetId(),
                                        sender.GetDescription());
    break;

  case TaskAttribute::HasCompleted:
    _taskStorage->UpdateTaskHasCompleted(sender.GetId(),
                                         sender.GetHasCompleted());
    break;
  }
}

void TaskController::GetAlltasks(
    std::function<void(std::vector<std::reference_wrapper<API::IToDoTask>>)>
        callback) {
  std::async(std::launch::async, [&]() {
    auto tasks = _taskStorage->GetAllTasks();
    for (auto const &refTask : tasks) {
      ToDoTask &task =
          (ToDoTask &)(std::unwrap_reference_t<API::IToDoTask &>(refTask));
      task.RegisterForPropertyChanged(
          [&](ToDoTask &sender, TaskAttribute changedAttribute) {
            OnTaskAttributesChanged(sender, changedAttribute);
          });
    }

    callback(std::vector<std::reference_wrapper<API::IToDoTask>>());
  });
}
} // namespace ToDo::Core
